design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,561 @@
+
+#### Joins
+
+Joins require the binary key of both sides of the join to match and both sides to be delivered to 
+the same ksqlDB node.  The former normally ensuring the latter, unless a custom partitioning 
+strategy has been used.
+
+The introduction of additional key formats means that while the deserialized key from both sides of 
+a join may match, the serialized binary data may differ if the key serialization format is different.
+To accommodate this, ksqlDB will automagically repartition one side of a join to match the key
+format of the other.
+
+Many joins require an implicit repartition of one or both sides to facilitate the join. In such 
+situations the change of key format can be performed in the same repartitioned step, avoiding any
+additional re-partitions. This means that joining sources with different key formats will only 
+require an implicit repartition to converge the key formats _if_ neither side is already being 
+repartitioned.
+
+Where one side must be repartitioned to correct the key format, choosing which side to reparation 
+can not be driven by the size of the data, as in a traditional database system, as the size of 
+the data is unknown, likely infinite. Ideally, for a streaming system it is the rate of change of 
+the data, i.e. the throughput, that would drive the choice. Unfortunately, this too can not be 
+known upfront.  For this reason, we propose repartitioning based on the order of sources within 
+the query, with the source on the _right_ being repartitioned.
+
+A benefit of making the choice order-based is that, once the rule is learned, users can predicate 
+and control which side is re-partitioned in some situations, i.e. stream-stream and table-table joins.
+
+Note: allowing users to freely switch left and right sources to control which side is repartitioned
+will work for all but left-outer joins. To support switching left-outer joins ksqlDB would need to
+support a right-outer join. The addition of this is deemed out of scope.
+
+Repartitioning the right side was chosen over the left, as it will mean stream-table joins will 
+repartition the table, which we propose will _generally_ see a lower throughput of updates to the 
+stream side. 
+
+Such repartitioning is possible and safe... ish, even for tables, because the logical key of the 
+data will not have changed, only the serialization format. This ensures the ordering of updates to 
+a specific key are maintained across the repartition. Of course, the repartitioning would introduce 
+cross-key out-of-order data, as the records are shuffled across partitions. That is to say that 
+even if the source partitions where correctly ordered by time, the re-partitioned partitions would 
+see out-of-order records, though per-key ordering would be maintained. Thus time-tracking 
+("stream-time"), grace-period and retention-time might be affected. However, this  phenomenon 
+already exists, and is deemed acceptable, for other implicit re-partitions.
+
```

Comment:
Add note clarifying that users can always repartition topics themselves before the join, in order to have full control over which sources are repartitioned (and that choosing the same sources ksqlDB would repartition and performing the repartitions upfront is equivalent from a resource-usage standpoint)? Or if not here, at least in the docs section so we don't forget to add the note later.